### PR TITLE
[Gradient Compression] Surface C++ comm hooks to Python API as built-in comm hooks

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46959 [Gradient Compression] Surface C++ comm hooks to Python API as built-in comm hooks**

1. Implemented the Pybind part.
2. In the reducer, once the builtin_comm_hook_type is set,  a c++ comm hook instance will be created in Reducer::autograd_hook.
3. Added unit tests for the builit-in comm hooks.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24471910](https://our.internmc.facebook.com/intern/diff/D24471910/)

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -1060,6 +1060,39 @@ def _register_comm_hook(self, state: object, hook: callable):
...
+        self, comm_hook_type: dist.BuiltinCommHookType
+    ):
+        r"""
+        Registers a built-in communication hook that specifies how DDP
+        aggregates gradients across multiple workers.
+
+        Arguments:
+            comm_hook_type (dist.BuiltinCommHookType): enum that corresponds
+                to a C++ comm hook implementation predefined in
+                torch/csrc/distributed/c10d/default_comm_hooks.h.
```
Comment (pritamdamania87) [inline torch/nn/parallel/distributed.py:1073]: This is implementation detail that we shouldn't mention in the docs here. You can add python docs to `BuiltinCommHookType` instead which would be available to python users.

File: test/distributed/test_c10d.py [hunk 2]
```diff
@@ -3474,18 +3489,45 @@ def allreduce_hook(state: object, bucket: dist._GradBucket) -> torch._C.Future:
...
+        """
+        This unit test verifies whether a built-in DDP communication hook that just calls
+        allreduce gives the same result result with the case of no hook registered.
+        """
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        # Get GPU model with the built-in allreduce communication hook.
+        gpu_model = self._gpu_model_with_builtin_ddp_comm_hook(
+            process_group, dist.BuiltinCommHookType.ALLREDUCE, gradient_as_bucket_view)
```
Comment (pritamdamania87) [inline test/distributed/test_c10d.py:3502]: Should we add a test for FP16_COMPRESS too?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 3]
```diff
@@ -1367,6 +1385,19 @@ void Reducer::register_comm_hook(std::unique_ptr<CommHookInterface> iface) {
...
+void Reducer::register_builtin_comm_hook(
+    c10d::BuiltinCommHookType comm_hook_type) {
+  TORCH_CHECK(
+      comm_hook_ == nullptr,
+      "register_builtin_comm_hook can only be called once.");
+  TORCH_CHECK(
+      replicas_.size() == 1,
+      "Communication hook does not support single-process multiple-device mode.");
+
+  builtin_comm_hook_type_ = comm_hook_type;
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/reducer.cpp:1398]: Why do we have this member field? We can just initialize `comm_hook_` with the appropriate builtin hook here right?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 4]
```diff
@@ -530,6 +531,23 @@ void Reducer::autograd_hook(VariableIndex index) {
...
+        break;
+      case c10d::BuiltinCommHookType::FP16_COMPRESS:
+        comm_hook_ =
+            std::make_unique<c10d::FP16CompressCommHook>(process_group_.get());
+        break;
+      default:
+        TORCH_WARN_ONCE(
+            "Unknown built-in DDP comm hook type is provided. No comm hook will be used.");
+    }
+  }
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/reducer.cpp:549]: This should be done in `register_builtin_comm_hook`

File: torch/distributed/algorithms/ddp_comm_hooks/__init__.py [hunk 5]
```diff
@@ -38,8 +39,23 @@ def register_ddp_comm_hook(
...
+        Registers the hooks of ``torch/csrc/distributed/c10d/default_comm_hooks.h``
+        to the DDP model. User can specify the type of hook as an enum
+        ``dist.BuiltinCommHookType`` type using ``comm_hook_type`` input.
+        Uses C++ comm hook implementations.
+
+        Example::
+            >>> register_builtin_ddp_comm_hook(dist.BuiltinCommHookType.FP16_COMPRESS, model)
+    """
+
+    model.reducer._register_builtin_comm_hook(comm_hook_type)
```
Comment (pritamdamania87) [inline torch/distributed/algorithms/ddp_comm_hooks/__init__.py:61]: Why do we have this additional API? Isn't the API in DDP sufficient?

File: torch/lib/c10d/ProcessGroupNCCL.hpp [hunk 6]
```diff
@@ -703,6 +727,6 @@ class ProcessGroupNCCL : public ProcessGroup {
   // by 1 when ncclGroupStart() is called and decreased by 1 when ncclGroupEnd()
   // is called.
   static thread_local uint64_t ncclActiveGroupCounter_;
-};
+}; // namespace c10d
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.hpp:730]: Did this come from clang-format? Doesn't seem right since the namespace actually ends below on line 732?

File: torch/nn/parallel/distributed.py [hunk 7]
```diff
@@ -1060,6 +1060,39 @@ def _register_comm_hook(self, state: object, hook: callable):
         self._check_comm_hook(hook)
         dist._register_comm_hook(self.reducer, state, hook)
 
+    def _register_builtin_comm_hook(
+        self, comm_hook_type: dist.BuiltinCommHookType
+    ):
+        r"""
+        Registers a built-in communication hook that specifies how DDP
+        aggregates gradients across multiple workers.
```
Comment (pritamdamania87) [inline torch/nn/parallel/distributed.py:1068]: Let's mention here that the purpose of the builtin hooks are to provide efficient C++ implementations for certain hooks which might not be as efficient if implemented in python using a python communication hook.

**Review by pritamdamania87 — COMMENTED** [review]
Can we also validate that this PR shows the same perf improvements on the HPC trainer as our earlier C++ hacks?

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (SciPioneer) [reply torch/nn/parallel/distributed.py:1073]: Thanks for catching this! Was not aware that this C++ code path is internal only.

File: test/distributed/test_c10d.py (hunk 2 above)
Comment (SciPioneer) [reply test/distributed/test_c10d.py:3502]: Didn't added such test because I found that the Python version of FP16_COMPRESS was not tested before.

Now lazily create a loop over both hooks, as they can provide the same test results.

Such style may not be sustainable with new hooks in the future, but I think we can use `torch.allclose` instead of `assertEqual` to allow approximation for any future hook that cannot support exact match.

File: torch/lib/c10d/ProcessGroupNCCL.hpp (hunk 6 above)
Comment (SciPioneer) [reply torch/lib/c10d/ProcessGroupNCCL.hpp:730]: I guess I somehow messed it up when resolving some merge conflicts.

File: torch/distributed/algorithms/ddp_comm_hooks/__init__.py (hunk 5 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/__init__.py:61]: You are totally right! It was created when I was thinking of passing the hook type as a pyobject earlier to be called from Python API. Forgot to delete it later. Good catch!

**Comment by pritamdamania87** [comment]
> It seems that this version is slightly (1%-2%) slower than the earlier C++ hacks. I suspect that the overheads are due to the usage of future.then() in our C++ comm hook.

Does this mean it is actually on par with the Python Hook or is the python hook slower than this hook?

**Review by pritamdamania87 — APPROVED** [review]
Looks good to me, although there are some merge conflicts in CI. Please fix those before landing.

**Comment by SciPioneer** [comment]
> Looks good to me, although there are some merge conflicts in CI. Please fix those before landing.

Need to run it on Ads 10x to verify. Unfortunately, that training command was somehow broken when I wrote this diff. Will verify it again later.

**Comment by SciPioneer** [comment]
> Looks good to me, although there are some merge conflicts in CI. Please fix those before landing.

Sure. I don't think I can land it before the fix.
